lib/optimize/ModuleConcatenationPlugin.js

```diff
@@ -0,0 +1,209 @@
+/*
+	MIT License http://www.opensource.org/licenses/mit-license.php
+	Author Tobias Koppers @sokra
+*/
+"use strict";
+
+const HarmonyImportDependency = require("../dependencies/HarmonyImportDependency");
+const ConcatenatedModule = require("./ConcatenatedModule");
+const HarmonyExportImportedSpecifierDependency = require("../dependencies/HarmonyExportImportedSpecifierDependency");
+
+class ModuleConcatenationPlugin {
+	constructor(options) {
+		if(typeof options !== "object") options = {};
+		this.options = options;
+	}
+
+	apply(compiler) {
+		compiler.plugin("compilation", (compilation, params) => {
+			params.normalModuleFactory.plugin("parser", (parser, parserOptions) => {
+				parser.plugin("call eval", () => {
+					parser.state.module.meta.hasEval = true;
+				});
+			});
+			compilation.plugin("optimize-chunk-modules", (chunks, modules) => {
+				chunks.forEach(chunk => {
+					const relevantModules = chunk.mapModules(m => m).filter(module => {
+						// Module must not be in other chunks
+						// TODO add an option to allow module to be in other entry points
+						if(module.getNumberOfChunks() !== 1)
+							return false;
+
+						// Because of variable renaming we can't use modules with eval
+						if(module.meta && module.meta.hasEval)
+							return false;
+
+						return true;
+					});
+					const possibleInners = new Set(relevantModules.filter(module => {
+						// Module must not be the entry points
+						if(chunk.entryModule === module)
+							return false;
+
+						// Exports must be known (and not dynamic)
+						if(!Array.isArray(module.providedExports))
+							return false;
+
+						// Using dependency variables is not possible as this wraps the code in a function
+						if(module.variables.length > 0)
```

Comment:
yes, basically if you define that using `$` should be `window.jQuery` then the module gets wrapped in

```
(function($){...}(window.jQuery))
```

(no guarantee for validity :P)